# Cleaning the Billionaires Statistics Data Set

## 1. Looking at basic data stats
## 2. Removing unnecessary columns and renaming columns
## 3. Looking for mixed data types and fixing data types
## 4. Looking for duplicates and inconsistencies




### 1. Looking at basic data stats

In [1]:
# Import Libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Creating path

path = r'C:\Users\Karen\Documents\Personal\Data Analytics\Career Foundry\Achievement 6\Data'

In [3]:
# Importing data set

df = pd.read_csv(os.path.join(path, 'Original Data', 'Billionaires Statistics Dataset.csv'))

In [4]:
# Removing the limit to columns displayed so I can see all the columns going forward

pd.options.display.max_columns = None

In [5]:
df.head()

rank  finalWorth               category                personName   age  \
0     1      211000       Fashion & Retail  Bernard Arnault & family  74.0   
1     2      180000             Automotive                 Elon Musk  51.0   
2     3      114000             Technology                Jeff Bezos  59.0   
3     4      107000             Technology             Larry Ellison  78.0   
4     5      106000  Finance & Investments            Warren Buffett  92.0   

         country    city              source             industries  \
0         France   Paris                LVMH       Fashion & Retail   
1  United States  Austin       Tesla, SpaceX             Automotive   
2  United States  Medina              Amazon             Technology   
3  United States   Lanai              Oracle             Technology   
4  United States   Omaha  Berkshire Hathaway  Finance & Investments   

  countryOfCitizenship                      organization  selfMade status  \
0               France  LVMH Moët Hennessy Louis Vuitton     False      U   
1        United States                             Tesla      True      D   
2        United States                            Amazon      True      D   
3        United States                            Oracle      True      U   
4        United States    Berkshire Hathaway Inc. (Cl A)      True      D   

  gender       birthDate lastName firstName                 title  \
0      M   3/5/1949 0:00  Arnault   Bernard      Chairman and CEO   
1      M  6/28/1971 0:00     Musk      Elon                   CEO   
2      M  1/12/1964 0:00    Bezos      Jeff  Chairman and Founder   
3      M  8/17/1944 0:00  Ellison     Larry       CTO and Founder   
4      M  8/30/1930 0:00  Buffett    Warren                   CEO   

            date       state residenceStateRegion  birthYear  birthMonth  \
0  4/4/2023 5:01         NaN                  NaN     1949.0         3.0   
1  4/4/2023 5:01       Texas                South     1971.0         6.0   
2  4/4/2023 5:01  Washington                 West     1964.0         1.0   
3  4/4/2023 5:01      Hawaii                 West     1944.0         8.0   
4  4/4/2023 5:01    Nebraska              Midwest     1930.0         8.0   

   birthDay  cpi_country  cpi_change_country           gdp_country  \
0       5.0       110.05                 1.1   $2,715,518,274,227    
1      28.0       117.24                 7.5  $21,427,700,000,000    
2      12.0       117.24                 7.5  $21,427,700,000,000    
3      17.0       117.24                 7.5  $21,427,700,000,000    
4      30.0       117.24                 7.5  $21,427,700,000,000    

   gross_tertiary_education_enrollment  \
0                                 65.6   
1                                 88.2   
2                                 88.2   
3                                 88.2   
4                                 88.2   

   gross_primary_education_enrollment_country  life_expectancy_country  \
0                                       102.5                     82.5   
1                                       101.8                     78.5   
2                                       101.8                     78.5   
3                                       101.8                     78.5   
4                                       101.8                     78.5   

   tax_revenue_country_country  total_tax_rate_country  population_country  \
0                         24.2                    60.7          67059887.0   
1                          9.6                    36.6         328239523.0   
2                          9.6                    36.6         328239523.0   
3                          9.6                    36.6         328239523.0   
4                          9.6                    36.6         328239523.0   

   latitude_country  longitude_country  
0         46.227638           2.213749  
1         37.090240         -95.712891  
2         37.090240         -95.712891  
3         37.090240 

In [6]:
df.shape

(2640, 35)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2640 entries, 0 to 2639
Data columns (total 35 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   rank                                        2640 non-null   int64  
 1   finalWorth                                  2640 non-null   int64  
 2   category                                    2640 non-null   object 
 3   personName                                  2640 non-null   object 
 4   age                                         2575 non-null   float64
 5   country                                     2602 non-null   object 
 6   city                                        2568 non-null   object 
 7   source                                      2640 non-null   object 
 8   industries                                  2640 non-null   object 
 9   countryOfCitizenship                        2640 non-null   object 
 10  organization

In [8]:
# Looking for basic stats for consistency check

df.describe()

rank     finalWorth          age    birthYear   birthMonth  \
count  2640.000000    2640.000000  2575.000000  2564.000000  2564.000000   
mean   1289.159091    4623.787879    65.140194  1957.183307     5.740250   
std     739.693726    9834.240939    13.258098    13.282516     3.710085   
min       1.000000    1000.000000    18.000000  1921.000000     1.000000   
25%     659.000000    1500.000000    56.000000  1948.000000     2.000000   
50%    1312.000000    2300.000000    65.000000  1957.000000     6.000000   
75%    1905.000000    4200.000000    75.000000  1966.000000     9.000000   
max    2540.000000  211000.000000   101.000000  2004.000000    12.000000   

          birthDay  cpi_country  cpi_change_country  \
count  2564.000000  2456.000000         2456.000000   
mean     12.099844   127.755204            4.364169   
std       9.918876    26.452951            3.623763   
min       1.000000    99.550000           -1.900000   
25%       1.000000   117.240000            1.700000   
50%      11.000000   117.240000            2.900000   
75%      21.000000   125.080000            7.500000   
max      31.000000   288.570000           53.500000   

       gross_tertiary_education_enrollment  \
count                          2458.000000   
mean                             67.225671   
std                              21.343426   
min                               4.000000   
25%                              50.600000   
50%                              65.600000   
75%                              88.200000   
max                             136.600000   

       gross_primary_education_enrollment_country  life_expectancy_country  \
count                                 2459.000000              2458.000000   
mean                                   102.858520                78.122823   
std                                      4.710977                 3.730099   
min                                     84.700000                54.300000   
25%                                    100.200000                77.000000   
50%                                    101.800000                78.500000   
75%                                    102.600000                80.900000   
max                                    142.100000                84.200000   

       tax_revenue_country_country  total_tax_rate_country  \
count                  2457.000000             2458.000000   
mean                     12.546235               43.963344   
std                       5.368625               12.145296   
min                       0.100000                9.900000   
25%                       9.600000               36.600000   
50%                       9.600000               41.200000   
75%                      12.800000               59.100000   
max                      37.200000              106.300000   

       population_country  latitude_country  longitude_country  
count        2.476000e+03       2476.000000        2476.000000  
mean         5.102053e+08         34.903592          12.583156  
std          5.542447e+08         17.003497          86.762989  
min          3.801900e+04        -40.900557        -106.346771  
25%          6.683440e+07         35.861660         -95.712891  
50%          3.282395e+08         37.090240          10.451526  
75%          1.366418e+09         40.463667         104.195397  
max          1.397715e+09         61.924110         174.885971

### The max tax rate is 106% - this is the tax rate assigned to Argentina. It is explained on the Kaggle discussion for this data set as follows: 
### source: https://idea.usaid.gov/cd/argentina/domestic-revenue-mobilization

### Total tax rate measures the amount of taxes and mandatory contributions payable by businesses after accounting for allowable deductions and exemptions as a share of commercial profits. Taxes withheld (such as personal income tax) or collected and remitted to tax authorities (such as value added taxes, sales taxes or goods and service taxes) are excluded.

### The education enrollment percentages are explained as follows: "All ratios are expressed as percentages and may exceed 100 because of early entry, repetition, and, for countries with almost universal education at a given level, whenever the actual age distribution of pupils extends beyond the official school ages." Source: https://www.who.int/data/gho/indicator-metadata-registry/imr-details/1142#:~:text=All%20ratios%20are%20expressed%20as,beyond%20the%20official%20school%20ages
----------------------------
### Remove these unnecessary columns: birthDate, date (not useful for analysis), status (contains undefined values), tax_revenue_country_country (not useful for analysis), source (not necessary for analysis because there are too many unique values and what is needed is the industry, not the specific business name or product), category (contains identical information as 'industry' field), lastName, firstName, birthYear, birthMonth, and birthDay because they are irrelevant to the analysis.
--------------------------------
### There appear to be missing values for a large number of columns. I'll address these individually below.

### 2. Removing Unnecessary Columns and Renaming Columns for Clarity

In [9]:
# Removing unnecessary columns

df=df.drop(columns = ['birthDate', 'date', 'status', 'tax_revenue_country_country', 'source', 'category', 
                      'lastName', 'firstName', 'birthYear', 'birthMonth', 'birthDay'])

In [10]:
# Checking that 12 columns were dropped

df.shape

(2640, 24)

In [11]:
# Investigating Missing Values

df.isnull().sum()

rank                                             0
finalWorth                                       0
personName                                       0
age                                             65
country                                         38
city                                            72
industries                                       0
countryOfCitizenship                             0
organization                                  2315
selfMade                                         0
gender                                           0
title                                         2301
state                                         1887
residenceStateRegion                          1893
cpi_country                                    184
cpi_change_country                             184
gdp_country                                    164
gross_tertiary_education_enrollment            182
gross_primary_education_enrollment_country     181
life_expectancy_country        

### There are too many missing values in the 'organization' and 'title' columns to be useful for analysis. So, these columns will be removed. 

### The 'state' and 'residenceStateRegion' are missing many values, however, the values it does contain apply to all billionaires in the United States and may be useful for analyzing that region. So, those will be left as they are. 

### The next largest number of values missing for a field is 184. This is less than 10% of the 2640 records and those fields can still be useful for analysis even with those missing values. So those will be left as they are.

In [12]:
# Removing columns with mostly missing values

df=df.drop(columns = ['organization', 'title'])

In [13]:
# Checking that 2 more columns were dropped

df.shape

(2640, 22)

In [14]:
# Renaming columns for clarity

df.rename(columns = {'finalWorth' : 'final_worth', 'personName' : 'name', 'countryOfCitizenship' : 'citizenship_country', 
                    'selfMade' : 'self_made', 'residenceStateRegion' : 'state_region', 
                     'gross_tertiary_education_enrollment' : 'tertiary_ed_enroll', 
                     'gross_primary_education_enrollment_country' : 'primary_ed_enroll'}, inplace = True)

### 3. Looking for mixed data types and fixing data types

In [15]:
# Looking for mixed data types

for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

country
city
state
state_region
gdp_country


### These columns with mixed data types are all strings with some data missing. I will not change the data types on these columns to make them completely strings because when doing that, the missing values no longer appear as missing when running this script: df.isnull().sum(). 
### The data that is there is the correct data type, with the exception of 'gdp_country'. That field needs to be converted to int64 or float64.

In [16]:
# Removing dollar signs from gdp_country column (received next line of code from ChatGPT)

df['gdp_country'] = df['gdp_country'].replace('[\$,]', '', regex=True).astype(float)

In [17]:
# Checking that dollar signs were removed and data type changed.

df['gdp_country'].head()

0    2.715518e+12
1    2.142770e+13
2    2.142770e+13
3    2.142770e+13
4    2.142770e+13
Name: gdp_country, dtype: float64

### 4. Looking for duplicates and inconsistencies

In [18]:
# Looking for duplicate records

df_dups = df[df.duplicated()]

In [19]:
df_dups

Empty DataFrame
Columns: [rank, final_worth, name, age, country, city, industries, citizenship_country, self_made, gender, state, state_region, cpi_country, cpi_change_country, gdp_country, tertiary_ed_enroll, primary_ed_enroll, life_expectancy_country, total_tax_rate_country, population_country, latitude_country, longitude_country]
Index: []

### There are no duplicate records

In [20]:
# Look at unique values in specific columns (received next line of code from ChatGPT)

unique_values = df['country'].unique()

In [21]:
# Print unique values

print(unique_values)

['France' 'United States' 'Mexico' 'India' 'Spain' 'China' 'Canada'
 'Germany' 'Switzerland' 'Belgium' 'Hong Kong' 'Austria' 'Japan'
 'United Kingdom' 'Australia' 'Indonesia' 'United Arab Emirates' 'Russia'
 'Chile' 'Monaco' 'Czech Republic' 'Sweden' nan 'Thailand' 'Uzbekistan'
 'Singapore' 'Nigeria' 'Israel' 'Italy' 'South Africa' 'Brazil' 'Malaysia'
 'South Korea' 'New Zealand' 'Philippines' 'Taiwan' 'Norway' 'Egypt'
 'Denmark' 'Eswatini (Swaziland)' 'Colombia' 'Netherlands' 'Poland'
 'Bahamas' 'Ukraine' 'Cayman Islands' 'Greece' 'Turkey' 'Argentina'
 'Georgia' 'Portugal' 'Kazakhstan' 'Algeria' 'Vietnam' 'Latvia' 'Finland'
 'Bermuda' 'Luxembourg' 'British Virgin Islands' 'Cambodia' 'Lebanon'
 'Oman' 'Ireland' 'Cyprus' 'Guernsey' 'Liechtenstein'
 'Turks and Caicos Islands' 'Romania' 'Qatar' 'Uruguay' 'Nepal' 'Slovakia'
 'Morocco' 'Hungary' 'Tanzania' 'Bahrain' 'Peru' 'Andorra' 'Armenia']


In [22]:
unique_values = df['city'].unique()

In [23]:
print(unique_values)

['Paris' 'Austin' 'Medina' 'Lanai' 'Omaha' 'New York' 'Mexico City'
 'Mumbai' 'Hunts Point' 'Palo Alto' 'La Coruna' 'Los Altos' 'Hangzhou'
 'Wichita' 'Bentonville' 'Fort Worth' 'Toronto' 'Ahmedabad' 'Hillsboro'
 'Beijing' 'Neckarsulm' 'Schindellegi' 'Brussels' 'The Plains' 'Jackson'
 nan 'Shenzhen' 'Las Vegas' 'Miami' 'Salzburg' 'Ningde' 'Tokyo' 'London'
 'Geneva' 'Shanghai' 'Kuenzelsau' 'Hong Kong' 'Haverford' 'East Setauket'
 'Bad Homburg' 'Perth' 'Delhi' 'Jakarta' 'Palm Beach' 'Ras Al Khaimah'
 'Frankfurt' 'Seattle' 'Kudus' 'Moscow' 'Foshan' 'Santiago' 'Pune' 'Laval'
 'Milton' 'Chicago' 'Osaka' 'Nashville' 'Greenwich' 'Monte Carlo'
 'Nanyang' 'Oklahoma City' 'Franklin' 'Indian Creek' 'Hisar'
 'Newport Beach' 'Eau Claire' 'Crans-Montana' 'Mulheim an der Ruhr'
 'Prague' 'Stockholm' 'Atherton' 'Zurich' 'San Jose' 'Bangkok' 'Baoding'
 'Tashkent' 'Singapore' 'Lagos' 'Changsha' 'Afton' 'Dallas' 'Tulsa'
 'Guangzhou' 'Sydney' 'Electra' 'Yinchuan' 'Chongqing' 'Naples' 'Dubai'
 'Tegernsee' 'C

In [24]:
unique_values = df['industries'].unique()
print(unique_values)

['Fashion & Retail' 'Automotive' 'Technology' 'Finance & Investments'
 'Media & Entertainment' 'Telecom' 'Diversified' 'Food & Beverage'
 'Logistics' 'Gambling & Casinos' 'Manufacturing' 'Real Estate'
 'Metals & Mining' 'Energy' 'Healthcare' 'Service'
 'Construction & Engineering' 'Sports']


In [25]:
unique_values = df['state'].unique()
print(unique_values)

[nan 'Texas' 'Washington' 'Hawaii' 'Nebraska' 'New York' 'California'
 'Kansas' 'Arkansas' 'Oregon' 'Virginia' 'Wyoming' 'Nevada' 'Florida'
 'Pennsylvania' 'Massachusetts' 'Illinois' 'Tennessee' 'Connecticut'
 'Oklahoma' 'Michigan' 'Wisconsin' 'Colorado' 'Indiana' 'New Hampshire'
 'Georgia' 'Missouri' 'New Jersey' 'North Carolina' 'Iowa' 'Kentucky'
 'Maryland' 'Montana' 'Ohio' 'South Carolina' 'Arizona' 'Louisiana' 'Utah'
 'Rhode Island' 'Idaho' 'U.S. Virgin Islands' 'Minnesota' 'Maine'
 'Mississippi' 'South Dakota' 'Alabama']


### There do not appear to be any inconsistencies in spellings or entry formats

In [26]:
# Exporting data frame as csv

df.to_csv(os.path.join(path, 'Prepared Data', 'billionaires_cleaned.csv'))